Add wxWebViewConfiguration

TcT2k opened the pull request:
In 40ff38b63b594e6a82bf6f7845e0ac2cb8ae2449 `wxWebView::GetNativeConfiguration()` was
added which isn't suitable if multiple `wxWebViewEdge` instances are created with
different options (as they must share the same options). To prevent such errors
the previous method is replaced by the new class `wxWebViewConfiguration` which
can be shared between various `wxWebView` instances.

This explicitly breaks API to improve usage (previous API wasn't available before 3.3.0).

In the future the new class could also enable a good way to wrap various common options/configurations available via native API.

TcT2k committed:
Add wxWebViewConfiguration

In 40ff38b63b594e6a82bf6f7845e0ac2cb8ae2449 wxWebView::GetNativeConfiguration() was
added which isn't suitable if multiple wxWebViewEdge instances are created with
different options (as they must share the same options). To prevent such errors
the previous method is replaced by the new class wxWebViewConfiguration which
can be shared between various wxWebView instances.

This explicitly breaks API to improve usage.

In the future the new class could also enable a good way to wrap various
common options/configurations available via native API.

vadz commented on include/wx/msw/webview_edge.h:
```diff
@@ -25,6 +25,8 @@ class WXDLLIMPEXP_WEBVIEW wxWebViewEdge : public wxWebView
 
     wxWebViewEdge();
 
+    wxWebViewEdge(std::shared_ptr<wxWebViewConfiguration> config);
```
```suggestion
    explicit wxWebViewEdge(std::shared_ptr<wxWebViewConfiguration> config);
```

vadz commented on include/wx/osx/webview_webkit.h:
```diff
@@ -28,27 +28,13 @@
 {
 public:
-    wxDECLARE_DYNAMIC_CLASS(wxWebViewWebKit);
+    wxWebViewWebKit(std::shared_ptr<wxWebViewConfiguration> config, wxWebViewWindowInfoWebKit* parentWindowInfo = nullptr);
```
```suggestion
    explicit wxWebViewWebKit(std::shared_ptr<wxWebViewConfiguration> config, wxWebViewWindowInfoWebKit* parentWindowInfo = nullptr);
```

vadz commented on include/wx/osx/webview_webkit.h:
```diff
@@ -135,15 +120,22 @@ class WXDLLIMPEXP_WEBVIEW wxWebViewWebKit : public wxWebView
                               const wxString& name = wxASCII_STR(wxWebViewNameStr)) override
-    { return new wxWebViewWebKit(parent, id, url, pos, size, style, name); }
+    {
+        auto webView = Create();
+        if (webView->Create(parent, id, url, pos, size, style, name))
```
I find it very confusing to see `Create()` called twice, typically this is an error in wx API (you're supposed to either use a non-default ctor or a default ctor and `Create()`). I wonder if there is still any way of renaming something to avoid calling both functions `Create()` or if it's too late?

TcT2k replied:
I agree it's confusing but the factory method is called `Create()` but i could of course use new here instead of calling `Create()` from the factory itself.

vadz commented on src/common/webview.cpp:
```diff
@@ -458,4 +478,15 @@ void wxWebView::InitFactoryMap()
+
+std::shared_ptr<wxWebViewConfiguration> wxWebViewFactory::CreateConfiguration()
+{
+    return std::shared_ptr<wxWebViewConfiguration>(new wxWebViewConfigurationDefault());
```
Should use `make_shared`.

vadz reviewed (commented):
My main question here is whether we really need to use `shared_ptr`. As it is now, it seems like we could just copy/pass by value to keep things simple. And if the configuration objects become too big to be copied all the time in the future (will they?) I'd prefer using `shared_ptr` in them internally to avoid copying rather than enforcing their use via shared pointers, which is rather confusing because it doesn't seem really obvious that we need shared ownership here.

If we do keep using shared pointers, I have a couple of nitpicks and would prefer to pass them by const reference instead of by value to the various functions and would also use `{}` instead of `nullptr` for an empty pointer. But, again, I'd very much prefer to not use them at all.

Would this be possible?

TcT2k commented:
> My main question here is whether we really need to use `shared_ptr`. As it is now, it seems like we could just copy/pass by value to keep things simple. And if the configuration objects become too big to be copied all the time in the future (will they?) I'd prefer using `shared_ptr` in them internally to avoid copying rather than enforcing their use via shared pointers, which is rather confusing because it doesn't seem really obvious that we need shared ownership here.
> 
> If we do keep using shared pointers, I have a couple of nitpicks and would prefer to pass them by const reference instead of by value to the various functions and would also use `{}` instead of `nullptr` for an empty pointer. But, again, I'd very much prefer to not use them at all.
> 
> Would this be possible?

Should be possible, just didn't think about doing it that way. Will certainly make some things easier.

vadz commented on include/wx/webview.h:
```diff
@@ -144,6 +144,27 @@ class WXDLLIMPEXP_WEBVIEW wxWebViewHandler
+    virtual void* GetNativeConfiguration() const { return nullptr; }
+};
+
+class WXDLLIMPEXP_WEBVIEW wxWebViewConfiguration
```
As written, this class shouldn't be DLL-exported as it only has inline functions. This would change, however, if `wxWebViewConfigurationImpl` becomes private as `GetNativeConfiguration()` couldn't be implemented inline then.

vadz commented on include/wx/webview.h:
```diff
@@ -144,6 +144,27 @@ class WXDLLIMPEXP_WEBVIEW wxWebViewHandler
+public:
+    explicit wxWebViewConfiguration(const wxString& backend, wxWebViewConfigurationImpl* impl)
+        : m_backend(backend), m_impl(impl) {}
+    virtual ~wxWebViewConfiguration() = default;
```
This class is not meant to be inherited from and so its dtor should _not_ be virtual.

vadz reviewed (commented):
Thanks a lot for the changes, this seems much cleaner to me now!

I'd just like to move the `Impl` class into private header if there is no reason to keep it public that I'm missing and then it would be perfect.

TcT2k committed:
Add wxWebViewConfiguration::SetDataPath()

While currently only implemented for the Edge backend GTK webkit could
also make use of such a setting.

Fixes: #19317

TcT2k committed:
Edge: Send an error event for various API errors

These errors are rare, but can happen. This way user code can react to
Previously they where just logged.

TcT2k commented:
I've moved `wxWebViewConfigurationImpl` to a private header and added some small additions.

TcT2k committed:
Edge: Hold environment in configuration

vadz commented:
Thanks again! Will merge in a moment.
